python/amici/sbml_import.py

```diff
@@ -955,42 +846,33 @@ def _process_rules(self) -> None:
 
                 for assignment in assignments:
                     assignments[assignment] = assignments[assignment].subs(
-                        variable, formula
+                        sym_id, formula
                     )
 
         # do this at the very end to ensure we have flattened all recursive
         # rules
-        for variable in assignments.keys():
-            self._replace_in_all_expressions(
-                sp.Symbol(variable, real=True),
-                assignments[variable]
-            )
+        for sym_id in assignments.keys():
+            self._replace_in_all_expressions(symbol_with_assumptions(sym_id),
+                                             assignments[sym_id])
 
-    def _process_volume_conversion(self) -> None:
+    @log_execution_time('processing SBML concentration conversion', logger)
+    def _process_concentration_conversion(self) -> None:
         """
-        Convert equations from amount to volume.
+        Convert species that only have amounts to concentration.
         """
-        compartments = self.species_compartment
-        for comp, vol in zip(self.compartment_symbols,
-                             self.compartment_volume):
-            if comp not in self.compartment_rate_rules:
-                compartments = compartments.subs(comp, vol)
-        for index, sunits in enumerate(self.species_has_only_substance_units):
-            if not sunits:
+        for species, definition in self.symbols[SymbolId.SPECIES].items():
+            if not definition['amount'] or 'compartment' not in definition:
                 continue
+            volume = self.compartments[definition['compartment']]
             self.flux_vector = \
-                self.flux_vector.subs(
-                    self.symbols['species']['identifier'][index],
-                    self.symbols['species']['identifier'][index]
-                    * compartments[index]
-                )
+                self.flux_vector.subs(species, species / volume)
```

Comment:
this is intended, now simulating species with "onlyHasSubstanceUnits" as amounts since concentrations aren't really well defined for them anyways and it simplifies import & validation.

Reply:
should document though!